### PR TITLE
userbrowse.py: implement path bar

### PR DESCRIPTION
**Pull request by mathiascode**
There are some issues left to sort out related to horizontal scrolling, but otherwise the path bar works fine.

The new design also allows for adding a filter bar later, similar to searches.

share

### CODE REVIEW

File: pynicotine/gtkgui/userbrowse.py [hunk 1]
```diff
@@ -447,8 +454,7 @@ def make_new_model(self, shares, private_shares=None):
         self.share_size = size + private_size
         self.num_folders = num_folders + num_private_folders
 
-        self.share_size_label.set_text(human_size(self.share_size))
-        self.num_folders_label.set_text(humanize(self.num_folders))
+        self.num_folders_label.set_text(f"{humanize(self.num_folders)}   /   {human_size(self.share_size)}")
```
Comment (slook) [inline pynicotine/gtkgui/userbrowse.py:457]: `f"{humanize(self.num_folders)} ({human_size(self.share_size)})"`

File: pynicotine/gtkgui/userbrowse.py [hunk 2]
```diff
@@ -603,11 +609,42 @@ def set_finished(self):
...
+    def populate_path_bar(self, folder_path=""):
+
+        for widget in list(self.path_bar):
+            self.path_bar.remove(widget)
+
+        for index, folder in enumerate(folder_path.split("\\")):
+            i_folder_path = "\\".join(folder_path.split("\\")[:index + 1])
+
+            if index:
+                label = Gtk.Label(label="/", visible=True)
```
Comment (slook) [inline pynicotine/gtkgui/userbrowse.py:621]: `"/"` -> `"\"`
Soulseek paths use backslashes.

Also, the horizontal padding seems a bit too spacious.

**Comment by slook** [comment]
The positioning of the "Expand / Collapse All" button looks like it should be a button which goes up a folder level.

Maybe consider combining it with the "Folder" label?

File: pynicotine/gtkgui/userbrowse.py (hunk 2 above)
Comment (mathiascode) [reply pynicotine/gtkgui/userbrowse.py:621]: I went with a backslash first, but @hboetes preferred a forward slash.

I can try reducing the padding slightly more, but I don't want to deviate from Nautilus' current design too much.

Comment (slook) [reply pynicotine/gtkgui/userbrowse.py:622]: In all other parts of the application, a backslash `\` is used in remote file paths, and on *nix a forward slash is used for local paths.

A forward slash here is technically incorrect on all systems, especially on Windows, where a forward slash is never seen in any path string.

**Comment by slook** [comment]
It's nice, but not really intuitive at first glance. It still looks like either a Download button or a Up a folder level button.

I'm beginning to wonder if the Expand/Collapse All button actually has any real value, or if it would be better to bury it in the Folder context menu or something like where the right click within empty space in the folder tree view.

**Comment by mathiascode** [comment]
Made some more changes. I think the only remaining issue is that the path bar doesn't always scroll all the way to the right in GTK 4.

**Comment by slook** [comment]
A horizontal border line is required at the bottom of the toolbar, or at the top of the tree views.

**Comment by mathiascode** [comment]
> A horizontal border line is required at the bottom of the toolbar, or at the top of the tree views.

I tried, but it felt a bit heavy with all the borders already present. I wonder if making the progress bar take up the full width of the tab would help correct the visual imbalance between the ends of the pane border.

**Comment by slook** [comment]
> making the progress bar take up the full width of the tab

The progress bar could ideally be moved into the toolbar to replace the path bar during loading operations, and then be invisible once loading is complete, because the path bar is not required during loading, and the progress bar widget doesn't serve any purpose after loading, other than just taking up space.

**Comment by mathiascode** [comment]
Alright, made some more changes.

**Comment by slook** [comment]
The progress bar doesn't appear at all on 3.3.0.dev5   •   Python 3.9.2   •   GTK 3.24.24

**Comment by mathiascode** [comment]
> The progress bar doesn't appear at all on GTK 3.24.24

How about https://github.com/nicotine-plus/nicotine-plus/pull/2661?

**Comment by slook** [comment]
Yes it works. Beautiful.

**Pull request merged**